pkg/ddc/vineyard/transform.go

```diff
@@ -0,0 +1,257 @@
+	}
+
+	if err := e.transformResourcesForWorker(runtime, value); err != nil {
+		return err
+	}
+
+	ports := e.transformWorkerPorts(runtime)
+	if len(ports) != 0 {
+		value.Worker.Ports = ports
+	}
+
+	err = e.transformWorkerVolumes(runtime, value)
+	if err != nil {
+		e.Log.Error(err, "failed to transform volumes for worker")
+	}
+
+	return
+}
+
+func (e *VineyardEngine) transformFuse(runtime *datav1alpha1.VineyardRuntime, value *Vineyard) {
+	value.Fuse = Fuse{}
+	image := runtime.Spec.Fuse.Image
+	imageTag := runtime.Spec.Fuse.ImageTag
+	imagePullPolicy := runtime.Spec.Fuse.ImagePullPolicy
+	value.Fuse.Image, value.Fuse.ImageTag, value.Fuse.ImagePullPolicy = e.parseFuseImage(image, imageTag, imagePullPolicy)
+
+	value.Fuse.CleanPolicy = runtime.Spec.Fuse.CleanPolicy
+
+	value.Fuse.NodeSelector = e.transformFuseNodeSelector(runtime)
+
+	value.Fuse.TargetPath = e.getMountPoint()
+	e.transformResourcesForFuse(runtime, value)
+}
+
+func (e *VineyardEngine) transformMasterSelector(runtime *datav1alpha1.VineyardRuntime) map[string]string {
+	properties := map[string]string{}
+	if runtime.Spec.Master.NodeSelector != nil {
+		properties = runtime.Spec.Master.NodeSelector
+	}
+	return properties
+}
+
+func (e *VineyardEngine) transformMasterPorts(runtime *datav1alpha1.VineyardRuntime) map[string]int {
+	ports := map[string]int{
+		MasterClientName: MasterClientPort,
+		MasterPeerName:   MasterPeerPort,
+	}
+	if len(runtime.Spec.Master.Ports) > 0 {
+		for key, value := range runtime.Spec.Master.Ports {
+			ports[key] = value
+		}
+	}
+	return ports
+}
+
+func (e *VineyardEngine) transformMasterOptions(runtime *datav1alpha1.VineyardRuntime) map[string]string {
+	options := map[string]string{
+		WorkerReserveMemory: DefaultWorkerReserveMemoryValue,
+		WorkerEtcdPrefix:    DefaultWorkerEtcdPrefixValue,
+	}
+	if len(runtime.Spec.Master.Options) > 0 {
+		for key, value := range runtime.Spec.Master.Options {
+			options[key] = value
+		}
+	}
+	return options
+}
+
+func (e *VineyardEngine) transformWorkerOptions(runtime *datav1alpha1.VineyardRuntime) map[string]string {
+	options := map[string]string{}
+	if len(runtime.Spec.Worker.Options) > 0 {
+		options = runtime.Spec.Worker.Options
+	}
+
+	return options
+}
+
+func (e *VineyardEngine) transformWorkerPorts(runtime *datav1alpha1.VineyardRuntime) map[string]int {
+	ports := map[string]int{
+		WorkerRPCName:      WorkerRPCPort,
+		WorkerExporterName: WorkerExporterPort,
+	}
+	if len(runtime.Spec.Worker.Ports) > 0 {
+		for key, value := range runtime.Spec.Worker.Ports {
+			ports[key] = value
+		}
+	}
+	return ports
+}
+
+func (e *VineyardEngine) transformFuseNodeSelector(runtime *datav1alpha1.VineyardRuntime) map[string]string {
+	nodeSelector := map[string]string{}
+	nodeSelector[e.getFuseLabelName()] = "true"
+	return nodeSelector
+}
+
+func (e *VineyardEngine) transformTieredStore(runtime *datav1alpha1.VineyardRuntime) TieredStore {
+	quota := resource.MustParse("4Gi")
```

Comment:
For vineyard users, How to set Vineyard's memory quota? And How to set spill quota and path in VineyardRuntime.spec.tieredstore?

Reply:
Thx. I wonder why there is a constant MEM level with 4Gi in `transformTieredStore`?

Reply:
If users doesn't set the any tieredstore fields in the `VineyardRuntime.spec`, the `4Gi` will be used as the default one.

Reply:
OK, but here's something tricky. Now the code firstly `value.TieredStore = e.transformTieredStore(runtime)`, and then calculates memory request of Vineyard workers according to runtimeInfo's `GetTieredStoreInfo()` in `e.transformResourcesForWorker()`. That makes an inconsistency when a default 4Gi MEM level is used. However, I do think it's a common problem for all the runtimes, so I think the code LGTM now. I will try to refactor the runtimeInfo code to make it clearer and more consistent. cc @cheyang

Reply:
Make sense to me. I have fixed it in the latest commit and it will raise an error if the tieredStore of vineyard runtime is empty.